lib/puppet/ssl/state_machine.rb

```diff
@@ -50,14 +50,28 @@ def initialize(machine)
     def next_state
       Puppet.debug("Loading CA certs")
 
+      force_crl_refresh = false
+
       cacerts = @cert_provider.load_cacerts
       if cacerts
         next_ctx = @ssl_provider.create_root_context(cacerts: cacerts, revocation: false)
+
+        now = Time.now
+        last_update = @cert_provider.ca_last_update
+        if needs_refresh?(now, last_update)
+          # set last updated time first, then make a best effort to refresh
+          @cert_provider.ca_last_update = now
```

Comment:
If there is an error acquiring the bundle, would we end up waiting another `:ca_refresh_interval` before attempting to update it again? Does it make sense to do this *after* we actually get the bundle?

Reply:
When I originally added this for CRLs, I was concerned about possibly getting into a state where all agents tried to refresh the CRL on every run. I was thinking something along the lines of we download the new CRL bundle, we try to save it to disk, something goes wrong and we never update the `last update` time. So then the next time we run, we download the new CRL bundle, etc. But now that the code has been out there for awhile in 7.x, I would be more inclined to do what you're suggesting.

For this PR, I'm thinking it would be good to keep the CA and CRL refresh behaviors consistent and then file another ticket to switch to "only update the last update time if were successful", so that agents are more likely to recover from failures like unreliable network, CA is restarted, etc.

Reply:
Filed as https://tickets.puppetlabs.com/browse/PUP-11869

Reply:
Oh dang it, I filed it as PUP-11870 :P